### DISCUSSION TITLE
Media queries on the web with SSR

### DISCUSSION THREAD

**Discussion by pierrezimmermannbam** [jpudysz/react-native-unistyles#116] [Q&A]
Hello @jpudysz ! Thanks again for this great lib, love using it so far!

I have a NextJS app using the pages router that does SSR. I tried using the experimental support for css media queries but it doesn't work for fontSize for some reason. From what I've investigated it looks like the generated classname does not match the one from RNWeb, but I guess this is expected because the feature is more of a preview (although if that's really the case I think it could be emphasized that this feature should not be used). Without using CSS for media queries, the behavior that I'm observing is that no style is applied at all. This is because on the server no breakpoint is set

```ts
if (!_common.isServer) {
  this.#breakpoint = this.getBreakpointFromScreenWidth(this.#screenWidth);
}
```

I wonder if it would be a good idea to allow users to configure the breakpoint to be used on the server. I was thinking of patching the lib to set the breakpoint on the server for now because I think it will improve DX. In my case all my font sizes are responsive and thus they all are at 14px by default until I resize the window which is not ideal, whereas by setting the server breakpoint to my desktop breakpoint it would feel like it works basically (it wouldn't if it rendered initially on mobile size). Maybe a better solution would be to update the breakpoint once we're client side but you'd probably get a weird effect though 

But maybe I'm completely missing something and we should not at all use media queries for components used on the web? I'd love to know how you manage these cases!

**Comment by jpudysz** [comment]
Hey man, I'm glad you love it! 💪

Generating CSS is currently in a preview stage due to some limitations:
- I have to depend on RNW and, for now, I need to implement some workaround methods (like generating identical classes) because the core is encapsulated. This should change in version 0.20, so I think that's the primary obstacle: [RNW Pull Request](https://github.com/necolas/react-native-web/pull/2508)
- Another problem is the lack of a CSS compiler! It will take some time to address all scenarios, including style specifity or properties like `textShadow` that depend on other properties. My plan is to release it with version 3.0, which will be dedicated to SSR/Web: [React Native UniStyles Discussion](https://github.com/jpudysz/react-native-unistyles/discussions/75)
- The third main reason is that I want to completely replace the StyleSheet from RNW with a custom solution

Regarding the second part of your comment, I'm not setting any breakpoints at the server because there's no `windowWidth` there. I can't predict your client window dimensions. To handle this scenario and possibly disable the experimental CSS, you can follow this approach: [Client-Side Configuration](https://reactnativeunistyles.vercel.app/reference/server-side-rendering/#client-side-configuration) (refer to the last example).

As for the use of media queries for components on the web, that's why I named it `experimentalCSSMediaQueries` 😇. It's a preview that will be fully addressed with the custom CSS compiler in version 3.0.0! 
Current implementation is too simple to make it production ready. It can work with `backgroundColor` or other stuff but eg. it can break with `fontSize`.

BTW here's a note indicating that it will work for basic cases only: [Web Support Reference](https://reactnativeunistyles.vercel.app/reference/web-support/).

I hope this makes sense!

**Reply by pierrezimmermannbam** [reply]
Thanks for the response, that's very clear!

If I understand correctly what the example implies, I would do something like this?

```tsx
const MyComponent = () => {
  const { styles, breakpoint } = useStyles(stylesheet);

  return <Text style={breakpoint === 'mobile' ? styles.mobile : styles.desktop} />;
};

const stylesheet = createStyleSheet(() => ({
  desktop: {
    fontSize: 20,
  },
  mobile: {
    fontSize: 16,
  },
}));
```

Also sorry I have another question, I don't really understand why the styles are not updated when the breakpoint is computed client side, my design will update only if I resize the window. Is there a reason for that?

**Reply by jpudysz** [reply]
Do you have an example? 

Please check example implementation with NextJS + Solito: https://github.com/jpudysz/react-native-unistyles-ssr-exmaple

**Reply by pierrezimmermannbam** [reply]
Yes I have tried using the example, if a style uses breakpoints and support for css media queries is disabled then on first page load the style is not applied at all even after hydration. It is applied for subsequent navigations though or after a resize of the window updates the breakpoint

**Reply by pierrezimmermannbam** [reply]
Oh ok I think I understood, in order for it to work the component must not be pre rendered on the server, otherwise I just get a mismatch error during hydration and the component doesn't update. This means that I cannot use breakpoints from unistyles for components that need to be rendered on the server. Sorry I'm just learning SSR so all these concepts are new to me, thanks again for your time!

**Reply by jpudysz** [reply]
Yeah, in order to make it 100% SSR I need custom CSS compiler. It will be possible with Unistyles in the near feature 😇

**Reply by pierrezimmermannbam** [reply]
Looking forward to it!
